processing: Add SFCGAL approximate medial axis processing

ptitjano opened the pull request:
## Description

This introduces a new processing based on SFCGAL backend: 

```
The Approximate Medial Axis algorithm generates a simplified skeleton of a shape by approximating its medial axis. 
The output is a collection of lines that follow the central structure of the shape. The result is a thin, stable set of curves that capture the main topology while ignoring noise.
This algorithm ignores the Z dimensions. If the geometry is 3D, the approximate medial axis will be calculated from its 2D projection
```

<img width="1920" height="1168" alt="image" src="https://github.com/user-attachments/assets/e9dc93fc-a55c-4ee9-9e92-3fd3448af458" />


Funded by Stadt Frankfurt am Main

ptitjano committed:
processing: Add approximate medial axis processing

SFCGAL backend is needed.

nyalldawson commented on src/analysis/CMakeLists.txt:
```diff
@@ -530,6 +530,13 @@ if (WITH_PDAL AND PDAL_2_5_OR_HIGHER)
   endif()
 endif()
 
+if (WITH_SFCGAL)
```
The correct way to handle this is to always expose the algorithm, but at runtime raise a processing exception of the required library is not available (with an explanatory message).


This let's users still work on models using these tools, and gives them a friendly explanation as to why the to won't work (as opposed to it just getting missing)

ptitjano replied:
I agree with that. However, I reproduced the existing logic of the PDAL processings. I QGIS is built without PDAL support the processing are not exposed.

ptitjano replied:
https://github.com/qgis/QGIS/issues/63106

ptitjano replied:
Done.

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,152 @@
+
+QString QgsApproximateMedialAxisAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Returns an approximate medial axis for the areal input based on its straight skeleton." );
```
```suggestion
  return QObject::tr( "Returns an approximate medial axis for a polygon layer input based on its straight skeleton." );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,152 @@
+
+Qgis::ProcessingAlgorithmFlags QgsApproximateMedialAxisAlgorithm::flags() const
+{
+  return QgsProcessingFeatureBasedAlgorithm::flags() | Qgis::ProcessingAlgorithmFlag::NoThreading;
```
Why this?

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,152 @@
+
+QStringList QgsApproximateMedialAxisAlgorithm::tags() const
+{
+  return QObject::tr( "medial,axis,create,lines" ).split( ',' );
```
```suggestion
  return QObject::tr( "medial,axis,create,lines,straight,skeleton" ).split( ',' );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,152 @@
+    {
+      inputSfcgalGeometry = QgsSfcgalGeometry( modifiedFeature.geometry() );
+    }
+    catch ( QgsSfcgalException const & )
+    {
+      feedback->reportError( QObject::tr( "Wrong geometry, cannot calculate approximate medial axis." ) );
+      modifiedFeature.clearGeometry();
```
```suggestion
    catch ( const QgsSfcgalException & e )
    {
      feedback->reportError( QObject::tr( "Cannot calculate approximate medial axis for feature %1: %2" ).arg( feature.id() ).arg( e.what() ) );
      modifiedFeature.clearGeometry();
```

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,152 @@
+        outputGeometry = QgsGeometry( outputSfcgalGeometry->asQgisGeometry() );
+        modifiedFeature.setGeometry( outputGeometry );
+      }
+      catch ( QgsSfcgalException const &medialAxisException )
+      {
+        feedback->reportError( medialAxisException.what() );
```
```suggestion
      catch ( const QgsSfcgalException &medialAxisException )
      {
        feedback->reportError( QObject::tr( "Cannot calculate approximate medial axis for feature %1: %2" ).arg( feature.id() ).arg( medialAxisException.what() ) );
```

nyalldawson commented on tests/src/analysis/testqgsprocessingsfcgalalgs.cpp:
```diff
@@ -0,0 +1,109 @@
+  QgsVectorLayer *outputLayer = qobject_cast<QgsVectorLayer *>( context->getMapLayer( results.value( QStringLiteral( "OUTPUT" ) ).toString() ) );
+  QVERIFY( outputLayer );
+  QVERIFY( outputLayer->isValid() );
+  QCOMPARE( outputLayer->featureCount(), 10 );
```
Let's actually test the results here too

ptitjano committed:
qgssfcgalgeometry: Add a default copy assignment operator

This fixes the following compilation error:
```
definition of implicit copy assignment operator for
'QgsSfcgalGeometry' is deprecated because it has a user-provided copy
constructor [-Werror,-Wdeprecated-copy-with-user-provided-copy] 89 |
QgsSfcgalGeometry( const QgsSfcgalGeometry &otherGeom ) SIP_THROW(
QgsSfcgalException );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmapproximatemedialaxis.cpp:
```diff
@@ -0,0 +1,167 @@
+QgsFeatureList QgsApproximateMedialAxisAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( context )
+  Q_UNUSED( feedback )
```
It's in use:
```suggestion
```

ptitjano replied:
Fixed. It's unused in the without SFCGAL case.

nyalldawson commented on tests/src/analysis/testqgsprocessingsfcgalalgs.cpp:
```diff
@@ -0,0 +1,157 @@
+  const QgsGeometry expectedGeom = openWktFile( "processing_medial_axis.wkt" );
+  QVERIFY( expectedGeom.wkbType() == Qgis::WkbType::MultiLineString );
+
+  QVERIFY( outputGeom.constGet()->fuzzyEqual( *expectedGeom.constGet() ) );
```
Let's compare via wkt instead, as that gives more useful debugging output if the tests fail then a simple boolean true/false result:

```suggestion
  QCOMPARE( outputGeom.constGet()->asWkt(4), expectedGeom.asWkt(4) ) );
```

(adjusting the precision as required)

ptitjano replied:
Done.
